File: Prime Sum/Prime Sum.java
import java.util.Scanner;
import java.util.Arrays;

public class Solution {
    static boolean prime(long[] primes, long n) {
        for(long i : primes) {
            if(i * i  > n) {
                break;
            }
            if(n % i == 0) {
                return false;
            }
        }
        return true;
    }
    
    static long[] sieve() {
        boolean[] prime = new boolean[999984];
        Arrays.fill(prime, true);
        for(int i = 2; i * i < 999984; i++) {
            if(prime[i]) {
                for(int j = i * i; j < 999984; j += i) {
                    prime[j] = false;
                }
            }
        }
        int size = 0;
        for(boolean i : prime) {
            if(i) {
                size++;
            }
        }
        long[] primes = new long[size];
        for(int i = 2, j = 0; i < 999984; i++) {
            if(prime[i]) {
                primes[j] = i;
                j++;
            }
        }
        return primes;
    }
    
    public static void main(String[] args) {
        long[] primes = sieve();
        Scanner scanner = new Scanner(System.in);
        for(int T = scanner.nextInt(); T > 0; T--) {
            long N = scanner.nextLong(), K = scanner.nextLong();
            if(K == 1) {
                if(prime(primes, N) && N != 1) {
                    System.out.println("Yes");
                } else {
                    System.out.println("No");
                }
            } else if(K == 2) {
                if(N % 2 == 0 && N != 2 || N > 3 && prime(primes, N - 2)) {
                    System.out.println("Yes");
                } else {
                    System.out.println("No");
                }
            } else {
                if(N < 2 * K) {
                    System.out.println("No");
                } else {
                    System.out.println("Yes");
                }
            }
        }
    }
}

File: Prime Sum/Prime Sum.cpp
#include <iostream>
#include <vector>
#pragma GCC target ("avx2")
#pragma GCC optimization ("O3")
#pragma GCC optimization ("unroll-loops")
using namespace std;

const long long SIZE = 999984;

bool prime(const vector<long long>& primes, long long n) {
    if (n <= 1) {
        return false;
    }
    for (long long p : primes) {
        if (p * p > n) {
            break;
        }
        if (n % p == 0) {
            return false;
        }
    }
    return true;
}

vector<long long> sieve() {
    vector<long long> prime;
    bool composite[SIZE] = {};
    for (long long i = 2; i < SIZE; i++) {
        if (!composite[i]) {
            prime.emplace_back(i);
        }
        for (long long j = 0; j < prime.size() && i * prime[j] < SIZE; j++) {
            composite[i * prime[j]] = true;
            if (i % prime[j] == 0) {
                break;
            }
        }
    }
    return prime;
}

int main() {
    vector<long long> primes = sieve();
    int t;
    cin >> t;
    for (long long n, k; cin >> n >> k; ) {
        if (k == 1) {
            if (prime(primes, n)) {
                cout << "Yes\n";
            } else {
                cout << "No\n";
            }
        } else if (k == 2) {
            if (n % 2 == 0 && n != 2 || prime(primes, n - 2)) {
                cout << "Yes\n";
            } else {
                cout << "No\n";
            }
        } else {
            if (n < 2 * k) {
                cout << "No\n";
            } else {
                cout << "Yes\n";
            }
        }
    }
    return 0;
}
